packages/python-yara-dex/PKGBUILD

```diff
@@ -0,0 +1,38 @@
+# This file is part of BlackArch Linux ( https://www.blackarch.org/ ).
+# See COPYING for license details.
+
+pkgname=python-yara-dex
+_pkgname=yara-python
+pkgver=1.0.3
+pkgrel=1
+pkgdesc='The Python interface for YARA.'
+url='https://github.com/MobSF/yara-python-dex'
+arch=('any')
+groups=('blackarch')
+license=('Apache')
+makedepends=('python-setuptools' 'git' 'glibc' 'python' 'yara')
+provides=('python-yara')
+conflicts=('python-yara' 'python-yara-rednaga')
```

Comment:
If they really conflicts (cf. https://wiki.archlinux.org/title/PKGBUILD#conflicts) have you check that other BA packages are using one of those dependencies? And if it's the case if they can move to `python-yara-dex`, else it will be problematic to have conflicting dependencies.

Reply:
They do conflict, yes, since APKID needs files from one of them, I haven't tried installing both to try and see if conflicting files are shown and pacman fails, but I believe that's what would happen. 

They could move to `python-yara-dex` but testing case by case would be needed, I have not tried to switch all packages that use `python-yara-rednaga` to `python-yara-dex` but according to this issue/comment : https://github.com/rednaga/APKiD/issues/318#issuecomment-1275072279
 
`python-yara-rednaga` is deprecated now, hence could explain the bug that was encountered with APKID.

But for that, another PR is needed, and I believe also that an issue should be created tracking packages that have moved on the new dependency, so that we can track the status in regards to all the pkgs that depends on `python-yara-rednaga` or `python-yara-1`

Reply:
What is the output of `grep -r python-yara packages`?

Reply:
Here is the output : 
```
packages/yeti/PKGBUILD:         'python-wtforms' 'python-xlsxwriter' 'python-yara' 'mongodb' 'redis'
packages/python-yara-dex/PKGBUILD:pkgname=python-yara-dex
packages/python-yara-dex/PKGBUILD:provides=('python-yara')
packages/python-yara-dex/PKGBUILD:conflicts=('python-yara' 'python-yara-rednaga')
packages/ssma/PKGBUILD:depends=('python' 'python-pefile' 'python-magic' 'python-yara-rednaga' 'ssdeep'
packages/quickscope/PKGBUILD:         'python-pefile' 'quark-engine' 'pyaxmlparser' 'python-yara'
packages/mobsf/PKGBUILD:         'python-requests' 'jre-openjdk' 'python-yara' 'sqlite' 'fontconfig'
packages/mobsf/pkg/mobsf/.BUILDINFO:installed = python-yara-dex-279.32b6a74-4-any
packages/mobsf/pkg/mobsf/.PKGINFO:depend = python-yara
packages/apkid/pkg/apkid/.BUILDINFO:installed = python-yara-dex-279.32b6a74-4-any
packages/apkid/pkg/apkid/.PKGINFO:depend = python-yara-dex
packages/apkid/pkg/apkid/.PKGINFO:makedepend = python-yara-dex
packages/apkid/PKGBUILD:depends=('python' 'python-argparse' 'python-yara-dex')
packages/apkid/PKGBUILD:makedepends=('python-setuptools' 'python-yara-dex')
packages/torcrawl/PKGBUILD:         'python-socks' 'python-urllib3' 'python-yara' 'tor')
packages/oledump/PKGBUILD:depends=('python' 'python-lxml' 'python-yara' 'python-olefile')
packages/pepper/PKGBUILD:depends=('python' 'python-pefile' 'python-yara' 'python-requests'
packages/python-plaso/PKGBUILD:         'python-xlsxwriter' 'python-yara' 'python-cryptography')
packages/peframe/PKGBUILD:         'python-yara' 'python-m2crypto' 'python-virustotal-api' 'python-magic'
packages/python-yara-rednaga/PKGBUILD:pkgname=python-yara-rednaga
packages/python-yara-rednaga/PKGBUILD:provides=('python-yara')
packages/python-yara-rednaga/PKGBUILD:conflicts=('python-yara')
packages/mkyara/PKGBUILD:depends=('python' 'python-capstone' 'python-yara' 'python-argparse')
packages/python-yaraprocessor/PKGBUILD:pkgbase=python-yaraprocessor
packages/python-yaraprocessor/PKGBUILD:pkgname=('python2-yaraprocessor' 'python-yaraprocessor')
packages/python-yaraprocessor/PKGBUILD:package_python-yaraprocessor() {
packages/stacs/PKGBUILD:depends=('python' 'python-click' 'python-pydantic' 'python-yara'
```

Reply:
- PKGBUILD [yara](https://github.com/archlinux/svntogit-community/blob/packages/yara/trunk/PKGBUILD) = not on pypi = github https://github.com/VirusTotal/yara
- PKGBUILD [python-yara](https://github.com/archlinux/svntogit-community/blob/packages/python-yara/trunk/PKGBUILD) = pypi [yara-python](https://pypi.org/project/yara-python/) = github https://github.com/VirusTotal/yara-python
- PKGBUILD [python-yara-rednaga](https://github.com/BlackArch/blackarch/blob/master/packages/python-yara-rednaga/PKGBUILD) = not on pypi  = github https://github.com/rednaga/yara-python-1
- PKGBUILD python-yara-dex = pypi [yara-python-dex](https://pypi.org/project/yara-python-dex) = github https://github.com/MobSF/yara-python-dex

1. To me it doesn't seem like there are conflicting so you should `provides` and `conflicts`. There are only `python-yara` and `python-yara-rednaga` that conflict each other but `python-yara-rednaga` is already flagged as so. I don't see `python-yara-dex` conflicting with neither of them.
2. To respect convention, this package should be named `python-yara-python-dex` and not `python-yara-dex`. I understand the suggested naming since the PKGBUILD `python-yara` should be named `python-yara-python`. It's common for python packages that include `python` in the name (eg. https://archlinux.org/packages/?q=python-python) to have several times the word `python` in the PKGBUILD name since the convention is `python-<python_package_name>`.

Reply:
Sorry about the delay for any updates.

About 2. : I understand, I've made the change, I'll push it tonight.

In respect to 1 :
So the pkg `python-yara-python-dex` should conflict only with itself ? And provides only `python-yara-python-dex` ? Or it should just conflict with `python-yara-rednaga` and provide it ?

Reply:
Sorry My sentence was missing the word `remove` . But thinking now, I just tested there was no conflict at the package installation but I didn't check for conflict at namespace import in python. So you should have to take a look at that.

Reply:
I've tried installing both packages `python-yara-rednaga`, `python-yara-python-dex` into a clean chroot and see, if they do collide, and there is a lib that seems to conflict : 
```
python-yara-rednaga: /usr/lib/python3.10/site-packages/yara.cpython-310-x86_64-linux-gnu.so exists in filesystem (owned by python-yara-python-dex)
```
Deleting this file, it seems that it is the only library/file that conflicts for both the packages.

As for the namespace Python, I'm not sure how I'd go around testing if there is indeed a conflict, if you have any advices on how to do that.

Looking into a Python REPL : 
```
>>> import yara
>>> yara.__file__   
'/usr/lib/python3.10/site-packages/yara.cpython-310-x86_64-linux-gnu.so'
```
This shows, that importing the `yara` module uses the library that both the packages share, I don't know if this could break any software, since the libraries are different.

Reply:
I believe then I'm done with this contribution, it was quite lengthy, if there is anything you believe I should improve on or better explain, then feel free to let me know, so this can be merged to fix, the issue with APKID.

Reply:
I was wrong to tell you to remove `provides` and `conflicts` as they are really conflicting as you demonstrated.
I don't think that we should move the dependencies of other packages to `python-yara-python-dex` as it will probably generate more issues.
Yes I'm sorry this review is length but it is a complex case and the outcome may impact many packages.
We have two choices here:

1. Keep apkid PKGBUILD as it is and use python-yara-python-dex but with the downside of having conflicting package so mobsf and the other packages you listed here https://github.com/BlackArch/blackarch/pull/3630#discussion_r1051488264 and that we try to avoid.
2. Totally change apkid PKGBUILD structure to use a virtual environment to avoid this dependency nightmare but also at a cost we also try to avoid because it makes much bigger package, more complex and harder to maintain.

I'm not sure myself of what the better options between those two. Any advice @noptrix?

Reply:
I understand, that it is a tricky case to solve due to the multiples yara packages variant, it seems to be a complex, I would lean towards solution 2 in the long run. 

However, I don't understand what you mean exactly by using a virtual env for APKID PKGBUILD, you mean that inside the PKGBUILD : 
1. A Python venv (virtual env) should be built in order to get `python-yara-python-dex` and build APKID using that, then install it.
2. Would APKID PKGBUILD install a virtual env (venv) and keep it for executing itself only with the specific dependency it needs ?

But I do agree that 2 would throw the complexity of maintaining that on the maintainers.

Reply:
APKID PKGBUILD using a standard jointscript PKGBUILD like https://github.com/BlackArch/blackarch-pkgbuilds/blob/master/PKGBUILD-python-standalone with a wrapper launching a virtual env. and The real whole installation happening in an `.install` script in the `post_install()` hook like that https://github.com/BlackArch/blackarch-pkgbuilds/blob/master/python-standalone.install.

Reply:
Hey, I've tried to do some modification, but I'm still not sure if I understood correctly what is the PKGBUILD python-standalone should contain along with the `.install`.

If I understand correctly, the PKGBUILD just builds a Python standalone virtual env and the `.install` called in post_install will install the dependency inside the virtual env using `pip` ?

But the way the PKGBUILD-python-standalone script example is made, it looks like it's for a python script for instance, here what we're packaging is a library for MobSF : 
![image](https://user-images.githubusercontent.com/14914796/232348112-831962b2-ce32-404f-a0ad-873418bca843.png)

Reply:
> If I understand correctly, the PKGBUILD just builds a Python standalone virtual env and the `.install` called in `post_install()` will install the dependency inside the virtual env using `pip` ?

Yes that's it.

apkid can be used as a CLI tool so this standalone setup would be right for that case. But in the case of apkid is used as a library (dependency) then it can't be installed in a standalone way. There are only two packages depending on apkid: mobsf  and quickscope. [Mobsf](https://github.com/BlackArch/blackarch/tree/master/packages/mobsf) is quick a mess too so a standalone packaging could apply to it too. Then if there is only quickscope left we could install it that way too.

Reply:
This looks like way too much work in order to just fix MobSF, I would rather use it with docker then, I might come back to try and fix this but fixing all this mess is way more than I can handle.
Thanks for your help in trying to move this PR towards a merge but there seems to be many PRs needed to have everything working "cleanly", I'd say

Reply:
Yes, most web platform tools are not easily packagable and often dependencies are not very well maintained so it's a nightmare to package.

Reply:
Thanks for the efforts.